Helm Charts Added

blacktig3r opened the pull request:
As per our discussion from [PR #348](https://github.com/onetimesecret/onetimesecret/pull/348)
I have created PR for helm-charts.

@venkatamutyala could you please review  and approve it?
Thanks,
Blacktig3r

venkatamutyala commented on values.yaml:
```diff
@@ -0,0 +1,16 @@
+redis:
+  image: redis:bookworm
+  service:
+    type: LoadBalancer
```
Shouldn't this be a ClusterIP?

blacktig3r replied:
yes..Updated in recent commit

venkatamutyala commented:
Do we want this chart to be stateless? I don't see a PVC definition. Am i missing something?

blacktig3r commented:
@venkatamutyala you are right...There should have been statefulset. I just added in recent commits. Please check.


```
root@YuvrajP-ub:~# kubectl get statefulset`
kubectNAME                  READY   AGE
onetimesecret-redis   3/3     2m38s


root@YuvrajP-ub:~# kubectl get pvc
NAME                                  STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
redis-storage-onetimesecret-redis-0   Bound    pvc-397bacd1-c524-4024-bfbb-eb41c5478b4f   1Gi        RWO            standard-rwo   2m47s
redis-storage-onetimesecret-redis-1   Bound    pvc-b298b6bd-6a4a-4535-ab95-3d136704ff9e   1Gi        RWO            standard-rwo   2m28s
redis-storage-onetimesecret-redis-2   Bound    pvc-706d5991-1a59-4d16-b0f5-2053c3b2f37d   1Gi        RWO            standard-rwo   2m10s


root@YuvrajP-ub:~# kubectl get pods
NAME                                           READY   STATUS    RESTARTS   AGE
nginx-1-6c5f6b76cf-l95rx                       1/1     Running   0          6d19h
nginx-1-6c5f6b76cf-spzcl                       1/1     Running   0          6d19h
nginx-1-6c5f6b76cf-t4b7m                       1/1     Running   0          6d19h
onetimesecret-onetimesecret-85b8bb77bb-5f8h5   1/1     Running   0          3m11s
onetimesecret-redis-0                          1/1     Running   0          3m11s
onetimesecret-redis-1                          1/1     Running   0          2m52s
onetimesecret-redis-2                          1/1     Running   0          2m33s
onetimesecret-redis-7cb5d478d-9g27r            1/1     Running   0          3m11s
```
`

venkatamutyala commented:
Have you tested this chart on a clean install? I see you have a redis deployment and a statefulset. have you considered using a redis helm chart as a sub chart?

blacktig3r commented:
Now I have added redis as a dependency chart and removed unwanted files.
Also I have tested chart on clean install. Pods are running fine but `onetimesecret`'s pod restart till `redis` pod become fully available.
 ```
NAME                                           READY   STATUS    RESTARTS      AGE
onetimesecret-onetimesecret-55f79fff5f-v92vr   1/1     Running   3 (42s ago)   67s
onetimesecret-redis-master-0                   1/1     Running   0             66s

```

delano commented:
Great work on this folks. Looks like it's close to being ready. 

Once we're happy let's update the readme and we can get it merged in.

blacktig3r commented:
sure @delano.
@venkatamutyala let me know if anything needs to be updated

venkatamutyala commented:
I'll give this a test tomorrow. Some immediate thoughts i had were whether or not we should use a bitnami helm chart or another vendor. I know bitnami is owned by VMWare and that Broadcom recently acquired VMWare. So basically i'm not sure what the future holds for that helm chart.

ArgoCD (another popular FOSS kubernetes project) uses this redis-ha as a sub chart:

https://github.com/argoproj/argo-helm/blob/main/charts/argo-cd/Chart.yaml#L20-L23


There are a few other things that we should possibly add like:
- the ability to use an existing ingress controller + annotation support for it.
- pinning to versions/shas for all the apps/images we use.

blacktig3r commented:
I have added a ingress configuration and updated the deployment templates to pin specific versions of the images used.
and About bitnami redis image, I think Broadcom's acquisition shouldn't affect the ongoing opensource projects.
But if you want I can test redis-ha.

venkatamutyala commented on values.yaml:
```diff
@@ -0,0 +1,30 @@
+    repository: onetimesecret/onetimesecret
+    tag: latest
+  service:
+    type: LoadBalancer
```
Not sure why it was a blocker but the onetimesecret pod wouldn't start in my environment until i set this to ClusterIP. I am running it locally via k3ds

venkatamutyala replied:
@blacktig3r any thoughts on this one?

venkatamutyala commented on values.yaml:
```diff
@@ -0,0 +1,30 @@
+
+ingress:
+  enabled: true
+  annotations: {}
```
Can you also add support for ingressClassName?

blacktig3r replied:
done..added in latest commit

venkatamutyala commented on values.yaml:
```diff
@@ -0,0 +1,30 @@
+onetimesecret:
+  image: 
+    repository: onetimesecret/onetimesecret
+    tag: latest
```
Can you set this to use: `{{ .Chart.AppVersion }}` ? This way it's defined in one spot.

blacktig3r replied:
I was not able to add template directly into values.yaml so I added it from template/onetimesecret-deployment.yaml which is going to overwrite the value with appVersion.

venkatamutyala commented on values.yaml:
```diff
@@ -0,0 +1,30 @@
+
+onetimesecret:
+  image: 
+    repository: onetimesecret/onetimesecret
```
Probably worth adding a registry option.

Example:

registry: "ghcr.io"  or "docker.io" as the default.

I think currently all the images are publishing to ghcr.io by default. We need to add docker.io to the main code repo but that shouldn't be a heavy lift.

blacktig3r replied:
added registry in latest commit

venkatamutyala commented on Chart.yaml:
```diff
@@ -0,0 +1,29 @@
+# incremented each time you make changes to the application. Versions are not expected to
+# follow Semantic Versioning. They should reflect the version the application is using.
+# It is recommended to use it with quotes.
+appVersion: "latest"
```
I think we should pin this to a particular version of the app. What do you think?

blacktig3r replied:
Yes..I think that's the best way we should follow. Pulling latest is not ideal

venkatamutyala commented:
@blacktig3r can you confirm what tag you are using in your test deployment? I'm getting a crashbackoff loop. I found this in my logs:


```
[Sat Jun  1 21:07:58 UTC 2024] INFO: Running entrypoint.sh...
Fetching gem metadata from https://rubygems.org/.
Resolving dependencies...
Installing bundler-graph 0.2.1
Bundle complete! 24 Gemfile dependencies, 58 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.
/app/lib/onetime.rb:9: warning: syslog was loaded from the standard library, but will no longer be part of the default gems since Ruby 3.4.0. Add syslog to your Gemfile or gemspec.
/usr/local/bundle/gems/sendgrid-ruby-6.7.0/lib/sendgrid/helpers/eventwebhook/eventwebhook.rb:1: warning: base64 was loaded from the standard library, but will no longer be part of the default gems since Ruby 3.4.0. Add base64 to your Gemfile or gemspec. Also contact author of sendgrid-ruby-6.7.0 to add base64 into its gemspec.
```

My ruby experience is a bit limited so nothing jumps out at me as a real error. I'm wondering if the latest is just not a great tag/image and i should be using something older

blacktig3r commented:
It restarts until redis become available, once redis pod is up onetimesecret's pod also become stable.
For me it took max 3 restarts(max 45 second)to become stable. I was checking if there is option to delay the launch of onetimesecret's pod until redis become available but no success.

venkatamutyala commented:
Huh. It's still erroring on my end. Let me take a look at my configs. It's possible my config for redis needs to be updated. I'm currently using the defaults.

venkatamutyala commented:
I think i just figured it out. I'll post my solution shortly. Need to step away for the next couple of hours.

venkatamutyala commented on values.yaml:
```diff
@@ -0,0 +1,30 @@
+    type: LoadBalancer
+    port: 3000
+  env:
+    REDIS_URL: "redis://onetimesecret-redis-master.default.svc.cluster.local:6379/0"
```
Maybe have the default be the namespace that it's installed in and if a value is provided have it use the value that the user provides? I'm thinking just commenting it out would be sufficient but let the user know it's available as an option easily. What do you think?

venkatamutyala commented on values.yaml:
```diff
@@ -0,0 +1,30 @@
+  env:
+    REDIS_URL: "redis://onetimesecret-redis-master.default.svc.cluster.local:6379/0"
+    COLONEL: "admin@example.com"
+    HOST: "onetimesecret-onetimesecret.default.svc.cluster.local:3000"
```
This value needs to be updated to the hostname or the correct namespace. I wonder if @delano can easily detect this in the code so we don't need it as a variable for the app.

blacktig3r replied:
I think I have found a way to update the URLs with correct namespaces. I'm overwriting the values dynamically from deployment.yaml template.

venkatamutyala commented on values.yaml:
```diff
@@ -0,0 +1,30 @@
+    COLONEL: "admin@example.com"
+    HOST: "onetimesecret-onetimesecret.default.svc.cluster.local:3000"
+    SSL: "false"  # Ensure this is a string
+
```
I think adding the remainder variables with example values would be helpful:


```
SMTP_HOST=
SMTP_PORT=
SMTP_USERNAME=
SMTP_PASSWORD=
FROM_EMAIL=
TO_EMAIL=
```

blacktig3r replied:
Adding these too

venkatamutyala reviewed (changes requested):
A few updates based on my recent experience deploying it.

venkatamutyala commented:
Left a few more comments just now. As we are getting towards the end, do you have a preferred place that we should publish this chart? I think we can do an OCI artifact with github packages/ghcr or we could do github pages too.

blacktig3r commented:
Hey @venkatamutyala , 
I have added changed which you suggested and they are working on my end.
I have added SMTP support as an example form so I commented those in files. Can you please take a look once?
and About publishing this chart, I have experience with github pages only...didn't worked on oci artifcat.

venkatamutyala commented on templates/onetimesecret-deployment.yaml:
```diff
@@ -0,0 +1,43 @@
+        image: "{{ .Values.onetimesecret.image.repository }}:{{ include "onetimesecret-chart.image.tag" . }}"
+        ports:
+        - containerPort: 3000
+        env:
```
So looking at the variables I think a better approach that I have seen is doing conditionals so that when something is defined it gets used otherwise it gets defaulted. Here is an example of what that might look like:

https://github.com/argoproj/argo-helm/blob/main/charts/argo-cd/templates/argocd-applicationset/deployment.yaml#L54-L56

In the case of REDIS_URL, if redis.enabled is false then it should probably take the value from the user as is.

In the case of SMTP host i'm thinking something similar. If it's defined take it from the user otherwise what are sane defaults to use? Is there a way to disable SMTP entirely or should the app error out? Is it required?

Let me know your thoughts. If you have seen a better practice i'm all for it. As for where the values should be commented out, I think having them in the values.yaml would be helpful so that the cosumer doesn't have to dig through the templates

venkatamutyala commented:
> Hey @venkatamutyala , I have added changed which you suggested and they are working on my end. I have added SMTP support as an example form so I commented those in files. Can you please take a look once? and About publishing this chart, I have experience with github pages only...didn't worked on oci artifcat.

Just left you a quick comment around this. I'll do my best to reply within 48 hours but give me 72 hours.

Also appreciate all the work you are doing on this.

blacktig3r commented:
> > Hey @venkatamutyala , I have added changed which you suggested and they are working on my end. I have added SMTP support as an example form so I commented those in files. Can you please take a look once? and About publishing this chart, I have experience with github pages only...didn't worked on oci artifcat.
> 
> Just left you a quick comment around this. I'll do my best to reply within 48 hours but give me 72 hours.
> 
> Also, I really do appreciate all the work you are doing on this.

Hi ,
Sorry, because of some emergency family work I was offline for two days. I'll try to take a look today.

venkatamutyala commented:
no worries! I'm still researching the publishing side too

blacktig3r commented:
Hi @venkatamutyala , I have added conditional approach for container port, REDIS_URL and SMTP_HOST.
I'm not sure about the enabling/disabling for SMTP, I think it is not possible. and If needed I think we should ask user inputs for smtp username and password as well and other smtp params we can keep as default. Let me know your thoughts

venkatamutyala commented on templates/onetimesecret-deployment.yaml:
```diff
@@ -0,0 +1,43 @@
+          value: "{{ include "onetimesecret-chart.onetimesecret.fullname" . }}{{ .Release.Namespace }}.svc.cluster.local:{{ include "onetimesecret-service.port" .}}"
+        - name: SSL
+          value: "{{ .Values.onetimesecret.env.SSL }}"
+### uncomment if you are adding smtp parameters in values.yaml 
```
I think this part should be uncommented @blacktig3r

venkatamutyala commented:
What's your availability to pair over a live call and hash through some of this PR together live? Can you suggest a couple times/dates you are available?

blacktig3r commented:
we can connect tomorrow(saturday) or Monday @8:30am Indian time zone.

venkatamutyala commented:
Yes, let's plan for Monday @8:30AM IST.

venkatamutyala commented on values.yaml:
```diff
@@ -0,0 +1,38 @@
+#   SMTP_USERNAME: "username"
+#   SMTP_PASSWORD: "password"
+#   FROM_EMAIL: "admin@example.com"
+#   TO_EMAIL: "recipent@example.com"
```
Per our discussion changing these examples to the same username@example.com ex: no-reply@example.com

blacktig3r replied:
Like we discussed I removed comments for SMTP and added a sample default values. Just instead of example.com I have added yourdomain.com.

fad3t commented on values.yaml:
```diff
@@ -0,0 +1,38 @@
+
+ingress:
+  enabled: true
+  ingressClassName: "ontimesecret-ingress"
```
Hi, I think you should leave this blank by default, so it uses the default cluster ingress class. And if needed, people can override it.

venkatamutyala replied:
@fad3t if you have one handy, can you share an example helm chart that does this already?

blacktig3r replied:
you are right, I removed it from here for default values and if user needed it they can now override.

blacktig3r commented:
@venkatamutyala I am having some trouble if I uses ClusterIP. I'll check in the morning.
Also I'll dig more onto redis part.
Thanks.

fad3t commented on values.yaml:
```diff
@@ -0,0 +1,35 @@
+  image: 
+    registry: docker.io
+    repository: onetimesecret/onetimesecret
+    tag: latest #will be overwritten by Chart.yaml appVersion
```
This value is useless in its current state, because you're using appVersion in any case.
I think it would be better to use `.Values.tag` and default to `.Chart.AppVersion` if its null.

blacktig3r replied:
added logic for `.Values.tag` and removed it from `values.yaml`
[f73950f](https://github.com/onetimesecret/helm-charts/pull/1/commits/f73950fd7972e539bc209b1499b9f6541abf7308)

fad3t replied:
Great, thanks. I think it's best to leave the default value in the `values.yaml` file, as a lot of user would go and read this value to see the possible configuration options.
You can either leave the value set to null (`tag: ~`) or just keep it commented.

blacktig3r replied:
I have kept tag commentd...also added note that 'this will override .Chart.AppVersion' .

blacktig3r commented:
ClusterIP is working now.
@venkatamutyala I have updated the Redis's chart version so that we will have the latest image of `Redis-7.2.5`.
Because like we discussed we have to consider the compatibility of `Valkey`. 
Let me know if I have missed any change and then we can move forward to publish it

fad3t commented on templates/_helpers.tpl:
```diff
@@ -0,0 +1,60 @@
+{{- end -}}
+
+{{- define "onetimesecret-chart.image.tag" -}}
+{{- if.Values.tag }}
+{{ .Values.tag }}
+{{- else }}
+{{ .Chart.AppVersion }}
+{{- end -}}
```
```suggestion
{{ default .Chart.AppVersion .Values.tag }}
```
FYI you can use this syntax to simplify (this applies to other helper functions above too).
BTW note you miss a space line 47.

blacktig3r replied:
Thanks for suggesting simplified way. I have applied it where it is possible in helper.

blacktig3r replied:
Thank you for suggesting simplified way for defaults. I have applied default for functions where it is possible.

blacktig3r commented:
> ClusterIP is working now. @venkatamutyala I have updated the Redis's chart version so that we will have the latest image of `Redis-7.2.5`. Because like we discussed we have to consider the compatibility of `Valkey`. Let me know if I have missed any change and then we can move forward to publish it

Also can we resolve our older "change" conversations If I have resolved those.

venkatamutyala commented:
> ClusterIP is working now. @venkatamutyala I have updated the Redis's chart version so that we will have the latest image of `Redis-7.2.5`. Because like we discussed we have to consider the compatibility of `Valkey`. Let me know if I have missed any change and then we can move forward to publish it

I think it's actually 7.2.4:

https://github.com/valkey-io/valkey/issues/43

7.2.5 from valkey appears to be their own flavor. Let me know if you found something that says otherwise.

venkatamutyala commented on values.yaml:
```diff
@@ -0,0 +1,35 @@
+  env:
+    COLONEL: "admin@example.com"
+    SSL: "false"  # Ensure this is a string
+    SMTP_HOST: "smtp.yourdomain.com" # Update your smtp configuration details
```
For consistency i would use "example.com" instead of "yourdomain.com". I believe example.com is owned by IANA while yourdomain.com is owned by an individual/org.

blacktig3r replied:
ohh..you are right. I have replaced it in latest commit

venkatamutyala reviewed (changes requested):
A few quick comments. Aside from that have you already tested the latest version of all of this?

venkatamutyala commented on values.yaml:
```diff
@@ -0,0 +1,35 @@
+  image: 
+    registry: docker.io
+    repository: onetimesecret/onetimesecret
+   #tag:latest  ##This will override the .Chart.AppVersion
```
There might be a better way to do this. Just looking at another helm chart:

https://github.com/argoproj/argo-helm/blob/main/charts/argo-cd/templates/argocd-repo-server/deployment.yaml#L72

https://github.com/argoproj/argo-helm/blob/main/charts/argo-cd/templates/_common.tpl#L37-L39

blacktig3r replied:
we have already defined same helper function as Argocd in helper. Just needed to remove comment for tag from `values.yaml` and keep it null.

venkatamutyala reviewed (changes requested):
One more comment around the tag

blacktig3r commented:
> > ClusterIP is working now. @venkatamutyala I have updated the Redis's chart version so that we will have the latest image of `Redis-7.2.5`. Because like we discussed we have to consider the compatibility of `Valkey`. Let me know if I have missed any change and then we can move forward to publish it
> 
> I think it's actually 7.2.4:
> 
> [valkey-io/valkey#43](https://github.com/valkey-io/valkey/issues/43)
> 
> 7.2.5 from valkey appears to be their own flavor. Let me know if you found something that says otherwise.

My bad..On valkey's official repo I saw the recent release was 7.2.5 that's why I was considering the same version for redis too. Corrected in latest commit

venkatamutyala reviewed (changes requested):
Can you squash all your commits?

blacktig3r commented:
> Can you squash all your commits?

Done. I have squashed all of my commits

venkatamutyala commented:
@blacktig3r are you able to merge on your own? If not, i can merge it in.
